implementations/rust/ockam/ockam_api/src/nodes/service.rs

```diff
@@ -657,18 +657,25 @@ impl NodeManagerWorker {
 
             // ==*== Inlets & Outlets ==*==
             (Get, ["node", "inlet"]) => {
-                let node_manager = self.node_manager.read().await;
-                self.get_inlets(req, &node_manager.registry).to_vec()?
+                if req.has_body() {
+                    self.show_inlet(req, dec).await?.to_vec()?
+                } else {
+                    let node_manager = self.node_manager.read().await;
+                    self.get_inlets(req, &node_manager.registry).to_vec()?
+                }
             }
             (Get, ["node", "outlet"]) => {
-                let node_manager = self.node_manager.read().await;
-                self.get_outlets(req, &node_manager.registry).to_vec()?
+                if req.has_body() {
+                    self.show_outlet(req, dec).await?.to_vec()?
+                } else {
+                    let node_manager = self.node_manager.read().await;
+                    self.get_outlets(req, &node_manager.registry).to_vec()?
+                }
```

Comment:
The endpoints should be:

- `node/inlet/{alias}`: show
- `node/inler`: list

And the same for `outlet`. Given that the `show` handlers only need the `alias` parameter, you can simplify them by removing the request body altogether.

Another little improvement you could do here: the `node_manager` lock is being held unnecessarily during the execution of the get/show function calls. This could be refactored as such:

```rust
let registry = {
    let node_manager = self.node_manager.read().await;
    &node_manager.registry
};
```

Reply:
Thanks for the suggestions, I think they made for way more straightforward endpoints. I refactored the `PortalAlias` struct out entirely from the `show` and `delete` commands using that first tip.

However, regarding the second improvement:
> Another little improvement you could do here: the `node_manager` lock [...]

After trying out several attempts at implementing it, not only did I get halted by the compiler with the same `borrow does not live long enough`, `lifetime ends here` errors, it also occurred to me this might possibly lead to a race condition if freed up before the request ends anyway - the registry could be modified before being read (or vice-versa), or modified improperly.

I tried:
- adding the block outside the `portals.rs` methods right after reading the request, replacing what's currently done;
- adding the `let registry { [...] }` block inside the `portals.rs` methods themselves;
- changing the `Registry` lifetime to `'static` (although I felt this would be a drawback in itself), and the fix you suggested

Let me know if you think it's still worth going down that route or if I misunderstood that second one! ✌️

Reply:
Oh sorry, you are right, I didn't see the problem when I wrote the comment. The issue is that you can't get a reference to the data that is held behind the lock when the lock has been dropped. The solution to making that approach work would be to clone the whole registry. To mitigate the cost of cloning the registry, you could refactor the `get_inlets/outlets` parameters to receive only the part of the registry that is needed.

> the registry could be modified before being read (or vice-versa), or modified improperly.

About this, note that we are only reading the registry, so a modification should not be possible or needed.

Reply:
> but the delete commands also use these registries and with write locks - if we had a delete write lock assume context after the cloning that part of the registry, wouldn't the output be inaccurate at this point (i.e. not correspond to the actual registry)?

Exactly, you can't apply the same trick to the endpoints that modify the registry. In those cases, it makes total sense to hold the lock until you are finished, so other concurrent calls to the NodeManager that affects the registry do not interfere with the current execution. As you can imagine, the problem here is the actual design of the registry: having all those sub-registries behind the same lock is not the most efficient idea because we shouldn't block modifying the tcp-connections registry when we know we only want to work with the tcp-outlets registry, for example. But right now is not critical either, so it's just fine.